rpc/client.go

```diff
@@ -234,19 +224,44 @@ func (c *Client) Close() {
 // runHeartbeat sends periodic heartbeats to client. Closes the
 // connection on error. Heartbeats are sent in an infinite loop until
 // an error is encountered.
-func (c *Client) runHeartbeat(stopper *util.Stopper) {
+func (c *Client) runHeartbeat(stopper *util.Stopper, reconnect func() error) {
 	if log.V(2) {
 		log.Infof("client %s starting heartbeat", c.Addr())
 	}
 
+	var failures int
+
 	for {
 		select {
+		case <-c.Closed:
+			return
 		case <-stopper.ShouldStop():
 			return
 		case <-time.After(heartbeatInterval):
 			if err := c.heartbeat(); err != nil {
-				log.Infof("client %s heartbeat failed: %v; recycling...", c.Addr(), err)
-				return
+				c.mu.Lock()
+				c.healthy = false
+				c.offset.Reset()
+				c.mu.Unlock()
+
+				failures++
+
+				log.Infof("client %s heartbeat failed %d times : %v; marked unhealthy", c.Addr(), failures, err)
+
+				if failures > 3 {
+					log.Infof("client %s reconnecting after %d heartbeat failures...", c.Addr(), failures)
+					if err := reconnect(); err != nil {
```

Comment:
So once an RPC client is created we'll keep trying to open connections to the remote node until the local node is restarted? This package is woefully under-documented. I think it would be good to call out some of the behavioral expectations in a comment at the top of the file.

@spencerkimball also needs to take a look at this. I'm familiar with the `rpc/codec` code, but this code is originally his.